system/jlib/jcontainerized.cpp

```diff
@@ -222,13 +222,36 @@ bool applyYaml(const char *componentName, const char *wuid, const char *job, con
 
     VStringBuffer args("\"--workunit=%s\"", wuid);
     args.append(" \"--k8sJob=true\"");
+    const char *runtimeImageVersion = getenv("baseImageVersion"); // runtime image version will equal base version unless changed dynamically below
     for (const auto &p: extraParams)
     {
-        if (hasPrefix(p.first.c_str(), "_HPCC_", false)) // job yaml substitution
+        // special handling _HPCC_JOB_VERSION_, not just a straight substitution
+        if (streq(p.first.c_str(), "_HPCC_JOB_VERSION_"))
+        {
+            // locates "image: <baseImageRootName>:<baseImageVersion>" and replaces with "image: <baseImageRootName>:<p.second>"
+            const char *baseImageRootName = getenv("baseImageRootName");
+            if (!isEmptyString(baseImageRootName) && !isEmptyString(runtimeImageVersion))
+            {
+                VStringBuffer oriImagePatternSpec("image: %s:%s", baseImageRootName, runtimeImageVersion);
+                VStringBuffer newImagePatternSpec("image: %s:%s", baseImageRootName, p.second.c_str());
+                jobYaml.replaceString(oriImagePatternSpec, newImagePatternSpec);
+                DBGLOG("Job image version changed from '%s' to '%s'", runtimeImageVersion, p.second.c_str());
+                runtimeImageVersion = p.second.c_str(); // used to substitute _HPCC_JOB_VERSION_ in jobYaml (used in runtimeImageVersion env variable)
+            }
+        }
+        else if (hasPrefix(p.first.c_str(), "_HPCC_", false)) // job yaml substitution
             jobYaml.replaceString(p.first.c_str(), p.second.c_str());
         else
             args.append(" \"--").append(p.first.c_str()).append('=').append(p.second.c_str()).append("\"");
     }
+    // always substitute _HPCC_JOB_VERSION_ - runtimeImageVersion is either the original baseImageVersion or the one from _HPCC_JOB_VERSION_ from above)
+    if (isEmptyString(runtimeImageVersion)) // if it were empty, it implies there's an incompatible helm chart/runtime image mismatch
+    {
+        // we replace _HPCC_JOB_VERSION_ (helm env runtimeImageVersion) with an empty string
+        // so that the engine knows it can't use it.
+        runtimeImageVersion = "";
```

Comment:
If runtimeImageVersion is blank I think that also implies there will not be any instances of  \_HPCC_JOB_VERSION\_ in the helm chart.
Also, it would generate:
```
- name: runtimeImageVersion
  value:
```
which I'm not sure is legal (missing value).  If you are replacing, possibly replace with "''"??

Reply:
the guard for blank was intended to catch incompatibility cases, where image had this functionality, but helm chart pre-dated them. In which case runtimeImageVersion will be blank. That means there will also be no '_HPCC_JOB_VERSION_' to replace either... since that can only exist if runtimeImageVersion(baseImageVersion) was present in the chart.

So I think this check/guard is pointless, instead it should be changed to only replace if runtimeImageVersion is non-empty.